state/state.go

```diff
@@ -42,9 +60,120 @@ type StatePersister interface {
 }
 
 // Locker is implemented to lock state during command execution.
-// The optional info parameter can be recorded with the lock, but the
-// implementation should not depend in its value.
+// The info parameter can be recorded with the lock, but the
+// implementation should not depend in its value. The string returned by Lock
+// is an ID corresponding to the lock acquired, and must be passed to Unlock to
+// ensure that the correct lock is being released.
+//
+// Lock and Unlock may return an error value of type LockError which in turn
+// can contain the LockInfo of a conflicting lock.
 type Locker interface {
-	Lock(info string) error
-	Unlock() error
+	Lock(info *LockInfo) (string, error)
+	Unlock(id string) error
+}
+
+// Generate a LockInfo structure, populating the required fields.
+func NewLockInfo() *LockInfo {
+	// this doesn't need to be cryptographically secure, just unique.
+	// Using math/rand alleviates the need to check handle the read error.
+	// Use a uuid format to match other IDs used throughout Terraform.
+	buf := make([]byte, 16)
+	rngSource.Read(buf)
+
+	id, err := uuid.FormatUUID(buf)
+	if err != nil {
+		// this of course shouldn't happen
+		panic(err)
+	}
+
+	// don't error out on user and hostname, as we don't require them
+	userName := ""
+	if userInfo, err := user.Current(); err == nil {
+		userName = userInfo.Username
+	}
+	host, _ := os.Hostname()
+
+	info := &LockInfo{
+		ID:      id,
+		Who:     fmt.Sprintf("%s@%s", userName, host),
+		Version: terraform.Version,
+		Created: time.Now().UTC(),
+	}
+	return info
+}
+
+// LockInfo stores lock metadata.
+//
+// Only Operation and Info are required to be set by the caller of Lock.
+type LockInfo struct {
+	// Unique ID for the lock. NewLockInfo provides a random ID, but this may
+	// be overridden by the lock implementation. The final value if ID will be
+	// returned by the call to Lock.
+	ID string
+
+	// Terraform operation, provided by the caller.
+	Operation string
+	// Extra information to store with the lock, provided by the caller.
+	Info string
+
+	// user@hostname when available
+	Who string
+	// Terraform version
+	Version string
+	// Time that the lock was taken.
+	Created time.Time
+
+	// Path to the state file when applicable. Set by the Lock implementation.
+	Path string
+}
+
+// Err returns the lock info formatted in an error
+func (l *LockInfo) Err() error {
+	return errors.New(l.String())
+}
+
+// Marshal returns a string json representation of the LockInfo
+func (l *LockInfo) Marshal() []byte {
+	js, err := json.Marshal(l)
+	if err != nil {
+		panic(err)
+	}
+	return js
+}
```

Comment:
I skipped the MarshalJSON so I could just call json.Marshal (otherwise it recurses), and since we control the data it shouldn't be able to error. I think only the encoding packages use TextMarshaler, and I wanted separate Text and JSON formats.

I can switch to MarshalJSON if you want, but I don't see a reason to use TextMarshaler.